### PR TITLE
Symbolication sample

### PR DESCRIPTION
**Pull request by olonho**
Converts backtrace like
```
Uncaught exception from Kotlin's main: kotlin.Error
        at 0   test.kexe                           0x0000000107505672 kfun:kotlin.Error.<init>()kotlin.Error + 50 [0x1074d7000]
        at 1   test.kexe                           0x00000001074df005 kfun:f(kotlin.Int) + 101 [0x1074d7000]
        at 2   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 3   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 4   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 5   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 6   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 7   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 8   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 9   test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 10  test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 11  test.kexe                           0x00000001074defd3 kfun:f(kotlin.Int) + 51 [0x1074d7000]
        at 12  test.kexe                           0x00000001074def9e kfun:main() + 14 [0x1074d7000]
        at 13  test.kexe                           0x00000001074def86 EntryPointSelector + 6 [0x1074d7000]
        at 14  test.kexe                           0x00000001074dee71 Konan_run_start + 113 [0x1074d7000]
        at 15  test.kexe                           0x00000001074dedeb Konan_main + 27 [0x1074d7000]
        at 16  libdyld.dylib                       0x00007fff6b53b08d start + 1 [0x7fff6b524000]
        at 17  ???                                 0x0000000000000001 0x0 + 1
```
to
```
Uncaught exception from Kotlin's main: kotlin.Error
        at 0   test.kexe                           kfun:kotlin.Error.<init>()kotlin.Error Exceptions.kt:11
        at 1   test.kexe                           kfun:f(kotlin.Int) test.kt:3
        at 2   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 3   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 4   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 5   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 6   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 7   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 8   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 9   test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 10  test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 11  test.kexe                           kfun:f(kotlin.Int) test.kt:1
        at 12  test.kexe                           kfun:main() test.kt:7
        at 13  test.kexe                           0x00000001074def86 EntryPointSelector + 6 [0x1074d7000]
        at 14  test.kexe                           0x00000001074dee71 Konan_run_start + 113 [0x1074d7000]
        at 15  test.kexe                           0x00000001074dedeb Konan_main + 27 [0x1074d7000]
        at 16  libdyld.dylib                       start AbstractList.kt:51
        at 17  ???                                 0x0000000000000001 0x0 + 1
```
Note that start is somewhat mishandled.

### CODE REVIEW

File: samples/symbolication/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Backtrace symbolization tool
+
+Produces more verbose backtrace symbolification with file name and line number.
+Only works on debug .dSYM binaries.
```
Comment (alexander-gorshenev) [inline samples/symbolication/README.md]: Symbolication, symbolization or symbolification? There are all three of them in the changes.

File: runtime/src/main/cpp/Porting.cpp [hunk 2]
```diff
@@ -341,6 +342,20 @@ long getpagesize() {
 #endif
 #endif
 
+
+intptr_t getImageBase(void* address) {
+  static uint64_t result = 0;
```
Comment (olonho) [inline runtime/src/main/cpp/Porting.cpp]: Incorrect caching.

File: runtime/src/main/cpp/Porting.h [hunk 3]
```diff
@@ -92,6 +92,9 @@ do {                          \
 } while(0)
 #endif
 
+// Computes program image load base address.
```
Comment (olonho) [inline runtime/src/main/cpp/Porting.h]: Incorrect docs.

File: samples/symbolication/src/nativeInterop/cinterop/coreSymbolication.def [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ unsigned long long length;
+} CSRange;
+
+typedef unsigned long long CSArchitecture;
+
+#define kCSNow LONG_MAX
+
+cpu_type_t CSArchitectureGetCurrent();
+
+CSSymbolicatorRef CSSymbolicatorCreateWithPid(pid_t pid);
```
Comment (olonho) [inline samples/symbolication/src/nativeInterop/cinterop/coreSymbolication.def]: Excessive API.

File: runtime/src/main/cpp/Natives.cpp [hunk 5]
```diff
@@ -47,6 +47,10 @@ OBJ_GETTER(Kotlin_getStackTraceStrings, KConstRef stackTrace) {
   RETURN_RESULT_OF(GetStackTraceStrings, stackTrace);
 }
 
+KLong Kotlin_getImageBase(KNativePtr address) {
+  return static_cast<KLong>(konan::getImageBase(address));
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Natives.cpp]: I guess this line sign-extends `address` to `KLong`. Is it correct on 32-bit platforms?

File: samples/symbolication/src/symbolicationMain/kotlin/main.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    val symbolicator = CSSymbolicatorCreateWithPathAndArchitecture(program, archToCpuName(arch))
+    if (CSIsNull(symbolicator)) throw Error("Cannot create \"$arch\" symbolicator for $program")
+    val owner = CSSymbolicatorGetSymbolOwner(symbolicator)
+    val imageName = CSSymbolOwnerGetName(owner)!!.toKString()
+    val imageBase = CSSymbolOwnerGetBaseAddress(owner)
+    // Format is like
+    // at 14  test.kexe                           0x0000000106a2e071 Konan_run_start + 113 [0x106a26000]
+    val matcher = "(at \\d+\\s+)(\\S+)(\\s+)0x([0-9a-f]+)\\s+([^+]+)\\+\\s\\d+\\s+\\[0x([0-9a-f]+)\\]".toRegex()
+    for (line in input) {
+        val match = matcher.find(line)
```
Comment (SvyatoslavScherbina) [inline samples/symbolication/src/symbolicationMain/kotlin/main.kt]: Why not `matchEntire`?

File: runtime/src/main/cpp/Natives.cpp (hunk 5 above)
Comment (olonho) [reply runtime/src/main/cpp/Natives.cpp]: Right.

File: runtime/src/main/kotlin/kotlin/Throwable.kt [hunk 7]
```diff
@@ -29,13 +30,18 @@ public open class Throwable(open val message: String?, open val cause: Throwable
...
+    public fun printStackTrace() {
         println(this.toString())
 
-        for (element in stackTraceStrings) {
-            println("        at " + element)
+        stackTraceStrings.forEachIndexed {
+            index, element ->
+            val base = getImageBase(stackTrace[index])
+            if (base != 0L)
+                println("        at $element [0x${base.toString(16)}]")
```
[the remark this answers is not in the document]
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/kotlin/Throwable.kt]: More specifically:
1. Get `CSSymbolRef` by name with `CSSymbolicatorGetSymbolWithNameAtTime` or something similar.
2. Get this symbol address with `CSSymbolGetRange(symbol).useContents { location }`.
3. Add offset.

The result appears to be a suitable replacement for current `imageAddress` value.

Motivation: such a symbolicator would work with released versions of Kotlin/Native.
Also stack traces would remain clean without image base.

[the remark this answers is not in the document]
Comment (olonho) [reply runtime/src/main/kotlin/kotlin/Throwable.kt]: Indeed, this made things way simpler.

File: samples/symbolication/src/symbolicationMain/kotlin/main.kt (hunk 6 above)
Comment (olonho) [reply samples/symbolication/src/symbolicationMain/kotlin/main.kt]: Either will work, I guess.

File: samples/symbolication/src/symbolicationMain/kotlin/main.kt [hunk 8]
```diff
@@ -0,0 +1,59 @@
+package symbolication
+
+import platform.darwin.*
+
+import kotlinx.cinterop.*
+import CoreSymbolication.*
+
+fun main(args: Array<String>) {
+    val program = if (args.size > 0 && args[0].isNotEmpty())
+        args[0] else throw Error("please specify .dSYM executable name as first argument")
```
[the remark this answers is not in the document]
Comment (vvlevchenko) [reply samples/symbolication/src/symbolicationMain/kotlin/main.kt]: yes

**Pull request merged**